android/jni/IndoorAtlas/IndoorAtlasLocationManagerJni.cpp

```diff
@@ -0,0 +1,18 @@
+// Copyright eeGeo Ltd (2012-2015), All Rights Reserved
+
+#include "IndoorAtlasLocationManagerJni.h"
+#include "IndoorAtlasLocationManager.h"
+#include "AndroidAppThreadAssertionMacros.h"
+
+JNIEXPORT void JNICALL Java_com_eegeo_indooratlas_IndoorAtlasJniMethods_OnLocationChanged(
+        JNIEnv* jenv, jobject obj,
+        jlong nativeObjectPtr,
+		jdouble latitude,
+		jdouble longitude)
+{
+    ASSERT_UI_THREAD
```

Comment:
This looks like a concurrency hazard to me:

1. IndoorAtlasLocationManager is being accessed via the native thread, I believe this to be the case because it's native -> java calls then enqueue runnables via runOnUiThread.

2. IndoorAtlasLocationManager is then directly accessed (OnUpdateLocation) from within the UI context.

Please publish a message, which is observed on the native context, using the message bus to more safely model this.